### DISCUSSION TITLE
writeContract with dynamic abi

### DISCUSSION THREAD

**Discussion by yoshiwarab** [wevm/viem#670] [Question] [answered]
Hi! I have a use case that requires loading abi's dynamically from a remote source and am running into type issues when calling `writeContract`. I know type inference won't work in this case but would expect the typing to allow this call regardless. Assuming `contractFunctionAbi` here is an object of type `AbiFunction`, this code

```
await walletClient.writeContract({
    address: contractAddress,
    abi: [contractFunctionAbi],
    functionName: contractFunctionABI.name,
    account: "0x0000",
  });
```

produces the ts error for functionName `Type 'string' is not assignable to type 'never'.ts(2322)
contract.d.ts(13, 5): The expected type comes from property 'functionName' which is declared here on type 'WriteContractParameters<[AbiFunction], string, Chain | undefined, Account | undefined, Chain | undefined>'`. Is this the expected behavior/is there some way around this or something else I'm missing. Thank you!

**Comment by jxom** [comment, marked as answer]
Should work if you cast it to `as Abi`: https://www.typescriptlang.org/play?target=99#code/JYWwDg9gTgLgBAbzgYygUwIYzQdQwG3zRgGF9g0A7GAGjgAsYYw6YBPMNOAQQCNg4AXzgAzKBBBwA5ADcKIKQChQkWIjggMwSpWJDR4ybPkB6ZPS2UAzksXII1+AHcCRUuSrwAvCnRZcrsRkFNQAFAiKcHAYyPYArtQAXNIADAAeKZlZ2Tm5efnZUjSRKBbayZraurQlMFAY1qowyYzMoQCUioKdii6EQR7UAHROUMDYJA51MTDhJRgAJgvoVlbJUukFW9tbRfP8yQjCGFY8-MVRIgnIMMAOAHIYIGjrEE66UADyIntRMgRxF5wFKULrtIA

**Reply by jxom** [reply]
You shouldn't need to pass `value` for nonpayable functions though. We need to fix up the type so it's not required for type `Abi` casted ABIs.

**Comment by yoshiwarab** [comment]
Ah, that's easy enough 😅, thank you!
